### ISSUE TITLE
Add support for `waitForTasksOutsideAngular` in protractor tests

### ISSUE THREAD

**Issue by mmalerba** [angular/components#18375]
Currently this has only been implemented in `TestbedHarnessEnvironment`, it should be implemented for `ProtractorHarnessEnvironment` too. I think we can reuse the same infrastructure we created to deal with this in unit tests, but just run it as a browser script in protractor. See https://github.com/angular/components/pull/18361 for some additional discussion

**Comment by LayZeeDK** [comment]
How about using `@sinonjs/fake-timers`?
https://github.com/sinonjs/fake-timers

**Comment by LayZeeDK** [comment]
Is this useful?

https://github.com/angular/blocking-proxy

**Comment by mmalerba** [comment]
Hmm, I'm not familiar with the blocking proxy, you're welcome to experiment with it and see if it helps. Like I mentioned above, I'm pretty sure we can just reuse the work we did for `TestbedHarnessEnvironment`, I just haven't had time to look into it yet.

**Comment by devversion** [comment]
The blocking proxy is definitely interesting, but probably not needed as we do not want to intercept all webdriver commands IIRC. At first glance, I'm thinking we will execute a client side script (similar to what bocking proxy does), but then only connect to it when we call `waitForTasksOutsideAngular` in the harness environment.

We should look into this, but also start running harness tests with the protractor test environment.

**Comment by tobiasschweizer** [comment]
I have a use case involving `MatAutocomplete`: the options are fetched from the server as the user types into the input field. 
Since this is async, the options will only be there once the answer from the server has been received and can be displayed in the template, see https://community.indepth.dev/t/create-a-component-harness-for-your-tests-with-angular-cdk/293/4?u=tobi.

Would calling `waitForTasksOutsideAngular` on `MatAutocompleteHarness` (when implemented) ensure that the async action has been successfully completed and that the `MatOptions` can be expected to be present?

**Comment by devversion** [comment]
@tobiasschweizer Yeah, that should do the trick. Assuming that the data fetching happens within `runOutsideAngular`. Is that the case here?

**Comment by tobiasschweizer** [comment]
> @tobiasschweizer Yeah, that should do the trick. Assuming that the data fetching happens within `runOutsideAngular`. Is that the case here?

No, it's not the case. So you suggest I should do the data fetching using [`runOutsideAngular`
](https://angular.io/api/core/NgZone#runoutsideangular) and re-enter via `run` once the data is present?

So then when using `waitForTasksOutsideAngular` in my custom harness, it would wait for the data fetching to be completed and then I can safely assume that `MatAutocomplete` contains the expected elements.

**Comment by LayZeeDK** [comment]
Work in progress: https://github.com/angular/components/pull/20351

**Comment by crisbeto** [comment]
Closing since Protractor is deprecated.
